### PR TITLE
Add distributed sync during wait_for_workers to avoid timeout for large checkpoints

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
Adds a distributed sync to the `RemoteUploaderDownloader.wait_for_workers` call so that the run does not NCCL timeout while uploading a large checkpoint at the end of a run.

Manual test:
`composer train.py yamls/finetune/mpt-7b_dolly_sft.yaml max_duration=2ba global_train_batch_size=4 device_train_microbatch_size=1 save_folder=oci://mosaicml-internal-checkpoints/daniel/checkpoints/{run_name} eval_subset_num_batches=2 eval_first=False optimizer.name=decoupled_lionw model.pretrained=False dist_timeout=20 loggers.wandb={} model.config_overrides.d_model=256 model.config_overrides.n_layers=4 run_name=test-run`
Before this PR, it hits a NCCL timeout during `FIT_END`. After this PR it finishes uploading the checkpoint, runs final eval, and exits cleanly.

# What issue(s) does this change relate to?
Closes [CO-2176](https://mosaicml.atlassian.net/browse/CO-2176)

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-2176]: https://mosaicml.atlassian.net/browse/CO-2176?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/loggers/remote_uploader_downloader.py [hunk 1]
```diff
@@ -463,9 +466,19 @@ def wait_for_workers(self):
...
+
         # Verify all tasks have been completed unless a worker threw an exception
-        while not self._file_upload_queue.empty() and self._exception_queue.empty():
+        all_ranks_upload_done_tensor = device.tensor_to_device(
+            torch.tensor([int(not self._file_upload_queue.empty() and self._exception_queue.empty())],
+                         dtype=torch.uint8))
+        dist.all_reduce(all_ranks_upload_done_tensor, reduce_operation='MAX')
+        while all_ranks_upload_done_tensor.item() == 1:
             time.sleep(0.2)
+            all_ranks_upload_done_tensor = device.tensor_to_device(
```
Comment (mvpatel2000) [inline composer/loggers/remote_uploader_downloader.py:477]: A dist reduce every 0.2s seems potentially too often? I'm not super sure though...

```suggestion
        while all_ranks_upload_done_tensor.item() == 1:
            time.sleep(2)
            all_ranks_upload_done_tensor = device.tensor_to_device(
```

**Review by mvpatel2000 — APPROVED** [review]
LGTM

File: composer/loggers/remote_uploader_downloader.py [hunk 2]
```diff
@@ -463,9 +466,19 @@ def wait_for_workers(self):
...
+        all_ranks_upload_done_tensor = device.tensor_to_device(
+            torch.tensor([int(not self._file_upload_queue.empty() and self._exception_queue.empty())],
+                         dtype=torch.uint8))
+        dist.all_reduce(all_ranks_upload_done_tensor, reduce_operation='MAX')
+        while all_ranks_upload_done_tensor.item() == 1:
+            time.sleep(2)
+            all_ranks_upload_done_tensor = device.tensor_to_device(
+                torch.tensor([int(not self._file_upload_queue.empty() and self._exception_queue.empty())],
+                             dtype=torch.uint8))
+            dist.all_reduce(all_ranks_upload_done_tensor, reduce_operation='MAX')
```
Comment (eracah) [inline composer/loggers/remote_uploader_downloader.py:480]: for readability maybe alias the boolean to a readable variable name
```suggestion
        uploads_still_not_done  = (all_ranks_upload_done_tensor.item() == 1)
        while uploads_still_not_done:
            time.sleep(2)
            all_ranks_upload_done_tensor = device.tensor_to_device(
                torch.tensor([int(not self._file_upload_queue.empty() and self._exception_queue.empty())],
                             dtype=torch.uint8))
            dist.all_reduce(all_ranks_upload_done_tensor, reduce_operation='MAX')
            uploads_still_not_done  = (all_ranks_upload_done_tensor.item() == 1)
```

**Review by eracah — APPROVED** [review]
nice LGTM besides some readability stuff. 

I wonder if we should do this same logic for exceptions. Right now exceptions are communicated intra-node with a multiprocessing queue, but I wonder if we should communicate them inter-node with torch.distributed + NCCL

**Pull request merged**